nixos/sshd: add enableRecommendedAlgorithms option

tomfitzhenry opened the pull request:
Prior to this commit, NixOS enabled a set of curated algorithms for OpenSSH.

This commit allows users to opt-out of this curation, and instead use the upstream algorithms. This also allows users to set Ciphers/KexAlgorithms/Macs themselves without lib.mkForce (and thus wield NixOS modules to build the list).

Tests have been added to ensure test this new option works.

I removed the part of the description that mentioned https://stribika.github.io/2015/01/04/secure-secure-shell.html and https://infosec.mozilla.org/guidelines/openssh#modern-openssh-67 since we have long since deviated from those.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

eclairevoyant commented:
@ofborg test openssh

tomfitzhenry commented:
> @tomfitzhenry Maybe try rebasing?

Done.

Interesting that this once passed, but now seems to cause a failure. I will diagnose the issue over the next few weeks.

Thanks for the reviews/investigations!

numinit commented:
`# The final segment in this IP is allocated according to the alphabetical order of machines in this test.`

Oh :sweat_smile: - that seems important.

tomfitzhenry commented:
@ofborg test openssh

tomfitzhenry commented:
> The final segment in this IP is allocated according to the alphabetical order of machines in this test.

Ah, I remember getting once confused by this, and adding a comment in 41911ed9d2ce9bc43fa32d796a62697bc62641b9. Alas, that comment wasn't enough, so I've updated this PR to rename the order-dependent machine to be alphabetically low (prefix with "aaa") so its IP ~won't change as new machines/tests are added.

tomfitzhenry committed:
tests/openssh: move order-dependent test to top

When working on nixosTests.openssh, I keep forgetting that there is
order-dependence, despite me adding a comment about this in commit
41911ed so let's solve ordering
once-and-for-all but making the order-dependent machine
predictably-first.

tomfitzhenry committed:
nixos/sshd: add enableRecommendedAlgorithms option

Prior to this commit, NixOS enabled a set of curated algorithms.

This commit allows users to opt-out of this curation, and instead use
the upstream algorithms. This also allows users to set
Ciphers/KexAlgorithms/Macs themselves without lib.mkForce (and thus
wield NixOS modules to build the list).

Tests have been added to ensure test this new option works.

tomfitzhenry commented:
nixosTests.openssh is failing on aarch64-linux. Moving to draft.

tomfitzhenry commented:
@ofborg test openssh

tomfitzhenry commented:
For anyone who wants to understand the difference in algorithms between `enableRecommendedAlgorithms = true` and `enableRecommendedAlgorithms = false`, here's how to find it:

```
$ nix build .#nixosTests.openssh.nodes.server.system.build.etc -o etc-nixos-recommended
$ nix build .#nixosTests.openssh.nodes.server-default-algorithms.system.build.etc -o etc-upstream
$ SSHD=$(nix build .#pkgs.openssh.out --no-link --print-out-paths)/bin/sshd
$ diff <($SSHD -f etc-upstream/etc/ssh/sshd_config -G | grep -Ei 'KexAlgorithms|Ciphers|MACs') <($SSHD -f etc-nixos-recommended/etc/ssh/sshd_config -G | grep -Ei 'KexAlgorithms|Ciphers|MACs')
1,3c1,3
< ciphers chacha20-poly1305@openssh.com,aes128-gcm@openssh.com,aes256-gcm@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr
< macs umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
< kexalgorithms mlkem768x25519-sha256,sntrup761x25519-sha512,sntrup761x25519-sha512@openssh.com,curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521
---
> ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
> macs hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,umac-128-etm@openssh.com
> kexalgorithms mlkem768x25519-sha256,sntrup761x25519-sha512,sntrup761x25519-sha512@openssh.com,curve25519-sha256,curve25519-sha256@libssh.org,diffie-hellman-group-exchange-sha256
```

Here's the output with a fancier diff tool (red is upstream, green is with `enableRecommendedAlgorithms`):

<img width="1781" height="370" alt="Image" src="https://github.com/user-attachments/assets/bab337bb-b4d3-4bd7-9129-0cef33c0c371" />
```

(That's `nix run .#pkgs.delta -- --hunk-header-style omit <($SSHD -f etc-upstream/etc/ssh/sshd_config -G | grep -Ei MACs) <($SSHD -f etc-nixos-recommended/etc/ssh/sshd_config -G | grep -Ei MACs)` for anyone that wants to copypaste it.)

tomfitzhenry commented on nixos/modules/services/networking/ssh/sshd.nix:
```diff
@@ -412,6 +412,16 @@ in
 
+      enableRecommendedAlgorithms = lib.mkOption {
+        type = lib.types.bool;
+        default = true;
```
Good idea. I've added this to the release notes. I've added a comment to this issue showing the differences.

numinit commented:
chromium file seems like a mistake?

tomfitzhenry commented:
> chromium file seems like a mistake?

Oops, fixed! That was left over from https://github.com/NixOS/nixpkgs/pull/500743#pullrequestreview-4072304283 (and me getting used to not having the git index with `jj`).

tomfitzhenry commented:
Rebased to master.

numinit reviewed (commented):
I am getting a reproducible failure:

`nix-build -A openssh.tests.nixosTest`

```
vm-test-run-openssh> server-lazy: must succeed: echo 'ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOSsTV6/GlHs9kBCq9gk2oFugw1Fq+8tsSEGcNiuPhnF root@client' > /root/.ssh/authorized_keys
vm-test-run-openssh> server-lazy: (finished: must succeed: echo 'ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOSsTV6/GlHs9kBCq9gk2oFugw1Fq+8tsSEGcNiuPhnF root@client' > /root/.ssh/authorized_keys, in 0.01)
vm-test-run-openssh> client: waiting for unit network.target
vm-test-run-openssh> client: (finished: waiting for unit network.target, in 0.04 seconds)
vm-test-run-openssh> client: must succeed: ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no server 'echo hello world' >&2
vm-test-run-openssh> client # ssh: connect to host server port 22: Network is unreachable
vm-test-run-openssh> client: output:
vm-test-run-openssh> !!! Test "manual-authkey" failed with error: "command `ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no server 'echo hello world' >&2` failed (exit code 255)"
vm-test-run-openssh> !!! Traceback (most recent call last):
vm-test-run-openssh> !!!   File "<string>", line 37, in <module>
vm-test-run-openssh> !!!     client.succeed(
vm-test-run-openssh> !!!
vm-test-run-openssh> !!! RequestedAssertionFailed: command `ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no server 'echo hello world' >&2` failed (exit code 255)
```

tomfitzhenry commented:
> I am getting a reproducible failure:
> 
> `nix-build -A openssh.tests.nixosTest`

nixosTests.openssh (and various other tests that rely on networking) are failing on master, which is why this PR's tests are failing. Tracked in https://github.com/NixOS/nixpkgs/issues/509254. I will rebase on the fix https://github.com/NixOS/nixpkgs/pull/509378 when it's merged.

numinit commented:
I had a suspicion it was due to that PR. Thanks for the bisect!

numinit reviewed (approved):
/nix/store/10v91g9wzazgf6vr3sc1d3gbql2ccnga-vm-test-run-borgbackup
/nix/store/5v80kklnxsf762q55lz6m73xc2fs9g11-vm-test-run-initrd-network-ssh
/nix/store/0ri5gc2l3kf881sq6dayrx61022dya85-vm-test-run-openssh
/nix/store/wwszaccq87jf62fylcnvkp0aqmdfm4za-openssh-test-10.2p1

LGTM now :)
